// ol/vectortile.js
import _ol_ from './index';
import _ol_Tile_ from './tile';
import _ol_TileState_ from './tilestate';

/**
 * @constructor
 * @extends {ol.Tile}
 * @param {ol.TileCoord} tileCoord Tile coordinate.
 * @param {ol.TileState} state State.
 * @param {string} src Data source url.
 * @param {ol.format.Feature} format Feature format.
 * @param {ol.TileLoadFunctionType} tileLoadFunction Tile load function.
 */
var _ol_VectorTile_ = function(tileCoord, state, src, format, tileLoadFunction) {

  _ol_Tile_.call(this, tileCoord, state);

  /**
   * @type {number}
   */
  this.consumers = 0;

  /**
   * @private
   * @type {ol.format.Feature}
   */
  this.format_ = format;

  /**
   * @private
   * @type {Array.<ol.Feature>}
   */
  this.features_ = null;

  /**
   * @private
   * @type {ol.FeatureLoader}
   */
  this.loader_;

  /**
   * Data projection
   * @private
   * @type {ol.proj.Projection}
   */
  this.projection_;

  this.replayGroups_ = {};

  /**
   * @private
   * @type {ol.TileLoadFunctionType}
   */
  this.tileLoadFunction_ = tileLoadFunction;

  /**
   * @private
   * @type {string}
   */
  this.url_ = src;

};

_ol_.inherits(_ol_VectorTile_, _ol_Tile_);


/**
 * @inheritDoc
 */
_ol_VectorTile_.prototype.disposeInternal = function() {
  this.features_ = null;
  this.replayGroups_ = {};
  this.state = _ol_TileState_.ABORT;
  this.changed();
  _ol_Tile_.prototype.disposeInternal.call(this);
};


/**
 * Get the feature format assigned for reading this tile's features.
 * @return {ol.format.Feature} Feature format.
 * @api
 */
_ol_VectorTile_.prototype.getFormat = function() {
  return this.format_;
};


/**
 * Get the features for this tile. Geometries will be in the projection returned
 * by {@link #getProjection}.
 * @return {Array.<ol.Feature|ol.render.Feature>} Features.
 * @api
 */
_ol_VectorTile_.prototype.getFeatures = function() {
  return this.features_;
};


/**
 * @inheritDoc
 */
_ol_VectorTile_.prototype.getKey = function() {
  return this.url_;
};


/**
 * Get the feature projection of features returned by {@link #getFeatures}.
 * @return {ol.proj.Projection} Feature projection.
 * @api
 */
_ol_VectorTile_.prototype.getProjection = function() {
  return this.projection_;
};


_ol_VectorTile_.prototype.getReplayGroup = function(key) {
  return this.replayGroups_[key];
};


/**
 * @inheritDoc
 */
_ol_VectorTile_.prototype.load = function() {
  if (this.state == _ol_TileState_.IDLE) {
    this.setState(_ol_TileState_.LOADING);
    this.tileLoadFunction_(this, this.url_);
    this.loader_(null, NaN, null);
  }
};


/**
 * Handler for successful tile load.
 * @param {Array.<ol.Feature>} features The loaded features.
 * @param {ol.proj.Projection} dataProjection Data projection.
 */
_ol_VectorTile_.prototype.onLoad_ = function(features, dataProjection) {
  this.setProjection(dataProjection);
  this.setFeatures(features);
};


/**
 * Handler for tile load errors.
 */
_ol_VectorTile_.prototype.onError_ = function() {
  this.setState(_ol_TileState_.ERROR);
};


/**
 * @param {Array.<ol.Feature>} features Features.
 * @api
 */
_ol_VectorTile_.prototype.setFeatures = function(features) {
  this.features_ = features;
  this.setState(_ol_TileState_.LOADED);
};


/**
 * Set the projection of the features that were added with {@link #setFeatures}.
 * @param {ol.proj.Projection} projection Feature projection.
 * @api
 */
_ol_VectorTile_.prototype.setProjection = function(projection) {
  this.projection_ = projection;
};


_ol_VectorTile_.prototype.setReplayGroup = function(key, replayGroup) {
  this.replayGroups_[key] = replayGroup;
};


/**
 * Set the feature loader for reading this tile's features.
 * @param {ol.FeatureLoader} loader Feature loader.
 * @api
 */
_ol_VectorTile_.prototype.setLoader = function(loader) {
  this.loader_ = loader;
};
export default _ol_VectorTile_;
